### PR TITLE
Sd 4598 enable workflow processing for oldvalue and ischanged

### PR DESCRIPTION
**Pull request by cosmicmeditator**

### CODE REVIEW

File: src/main/java/com/kylas/sales/workflow/domain/processor/WorkflowProcessor.java [hunk 1]
```diff
@@ -88,7 +91,30 @@ public void process(EntityEvent event) {
...
+        return satisfies(expression.getOperand1(), event) && satisfies(expression.getOperand2(), event);
+      } else if (expression.getOperator().equals(Operator.OR)) {
+        return satisfies(expression.getOperand1(), event) || satisfies(expression.getOperand2(), event);
+      }
+    } else if (TriggerType.IS_CHANGED.equals(expression.getTriggerOn())) {
+      return satisfiesValueIsChanged(event, expression);
+    }
+    return expression.getTriggerOn().equals(TriggerType.NEW_VALUE)
+        ? conditionFacade.satisfies(expression, event.getEntity())
+        : conditionFacade.satisfies(expression, event.getOldEntity());
```
Comment (abhishek-bhuskute) [inline src/main/java/com/kylas/sales/workflow/domain/processor/WorkflowProcessor.java:109]: Will we ever reach the last return statement?

File: src/main/java/com/kylas/sales/workflow/domain/processor/WorkflowProcessor.java [hunk 2]
```diff
@@ -88,7 +91,30 @@ public void process(EntityEvent event) {
...
+  }
+
+  private boolean satisfiesValueIsChanged(EntityEvent event, ConditionExpression expression) {
+    EvaluationContext context = SimpleEvaluationContext.forReadWriteDataBinding().build();
+    ExpressionParser parser = new SpelExpressionParser();
+    var oldPropertyName = parser.parseExpression(expression.getName()).getValue(event.getOldEntity());
+    var newPropertyName = parser.parseExpression(expression.getName()).getValue(event.getEntity());
+    return !newPropertyName.equals(oldPropertyName);
```
Comment (abhishek-bhuskute) [inline src/main/java/com/kylas/sales/workflow/domain/processor/WorkflowProcessor.java:117]: Does this handle nested null value? What happens if you try to fetch property `salutation.id` and `salutation` itself is null?
Extract a method out of the handled code fragment.

File: src/test/resources/contracts/mq/events/lead-created-triggering-conditional-workflows-old-value-multiple-conditions.json [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+  },
+  "metadata": {
+    "entityAction": "CREATED",
+    "tenantId": 99,
+    "userId": 12,
+    "entityType": "LEAD",
+    "workflowId": null,
+    "executedWorkflows": []
+  }
+}
```
Comment (abhishek-bhuskute) [inline src/test/resources/contracts/mq/events/lead-created-triggering-conditional-workflows-old-value-multiple-conditions.json:20]: Shouldn't we refer to the same file if their content is exactly same-
- lead-created-triggering-conditional-workflows--multiple-conditions.json
- lead-created-triggering-conditional-workflows-ischanged-multiple-conditions.json
- lead-created-triggering-conditional-workflows-old-value-multiple-conditions.json

**Review by abhishek-bhuskute — CHANGES_REQUESTED** [review]
Needs some fixes. Also, check if this breaks existing functionality.

File: src/test/resources/contracts/mq/events/lead-created-triggering-conditional-workflows-old-value-multiple-conditions.json (hunk 3 above)
Comment (abhishek-bhuskute) [reply src/test/resources/contracts/mq/events/lead-created-triggering-conditional-workflows-old-value-multiple-conditions.json:20]: This hasn't been addressed yet.

**Review by abhishek-bhuskute — CHANGES_REQUESTED** [review]
I still see 3 new identical JSON files.

**Review by abhishek-bhuskute — APPROVED** [review]

**Review by bhaskarpawar123 — APPROVED** [review]

**Pull request merged**